File: intern/67.add-binary.js
/*
 * @lc app=leetcode id=67 lang=javascript
 *
 * [67] Add Binary
 */

// @lc code=start
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
const addBinary = function (a, b) {
    const LENGTH = Math.max(a.length, b.length);
    const ZERO_ASCII = '0'.charCodeAt(0);
    a = a.padStart(LENGTH, '0');
    b = b.padStart(LENGTH, '0');
    const result = [];
    let carry = 0;
    for (let i = LENGTH - 1; i >= 0; i--) {
        const aBit = a.charCodeAt(i) - ZERO_ASCII;
        const bBit = b.charCodeAt(i) - ZERO_ASCII;

        switch (aBit + bBit + carry) {
            case 3: {
                result.push('1');
                carry = 1;
                break;
            }
            case 2: {
                result.push('0');
                carry = 1;
                break;
            }
            case 1: {
                result.push('1');
                carry = 0;
                break;
            }
            case 0: {
                result.push('0');
                carry = 0;
                break;
            }
        }
    }

    if (carry === 1) {
        result.push('1');
    }

    result.reverse();
    return result.join('');
};
// @lc code=end

File: src/2026/67.add-binary.cpp
/*
 * @lc app=leetcode id=67 lang=cpp
 *
 * [67] Add Binary
 */

#include <string>

// @lc code=start
class Solution {
 public:
  std::string addBinary(std::string a, std::string b) {
    const std::string& shorter = a.size() < b.size() ? a : b;
    std::string& longer = a.size() < b.size() ? b : a;

    auto shorter_reverse_it = shorter.crbegin();
    auto longer_reverse_it = longer.rbegin();

    int carry = 0;
    while (shorter_reverse_it != shorter.rend()) {
      const int sum =
          (*shorter_reverse_it - '0') + (*longer_reverse_it - '0') + carry;
      *longer_reverse_it = sum % 2 + '0';
      carry = (sum - sum % 2) / 2;
      ++shorter_reverse_it;
      ++longer_reverse_it;
    }

    while (carry > 0 && longer_reverse_it != longer.rend()) {
      const int sum = (*longer_reverse_it - '0') + carry;
      *longer_reverse_it = sum % 2 + '0';
      carry = (sum - sum % 2) / 2;
      ++longer_reverse_it;
    }

    while (carry > 0) {
      longer.insert(0, "1");
      carry--;
      carry /= 2;
    }

    return longer;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.addBinary("1010", "1011");
}